web/main/models.py

```diff
@@ -1429,6 +1429,29 @@ class Meta:
         unique_together = (("user", "casebook"),)
 
 
+class CasebookTag(models.Model):
+    casebook = models.ForeignKey(
+        "Casebook",
+        on_delete=models.CASCADE,
+    )
+    tag = models.ForeignKey(
+        "CasebookTagName",
+        on_delete=models.CASCADE,
+    )
+
+    class Meta:
+        unique_together = (("tag", "casebook"),)
+
+
+class CasebookTagName(models.Model):
```

Comment:
I was a little confused briefly by the model naming here—I think maybe just call this a `Tag` and call the other one `CasebookTag`, as you have. The convention for many-to-many through models is to name the thing the concatenation of the two model names, so `Casebook + Tag` makes sense but then I had trouble parsing what I assume is  `CasebookTag [a type of tag] Name`.  Just calling it `Tag` resolves the issue, plus it leaves open the possibility of using it for sub-casebook level tagging.

Reply:
Yeah I was really unsure about the naming --- I tried a few different variants all of which were equally confusing. `Tag`/`CasebookTag` makes sense; it is also more clear to me now from @cath9's outline that we are hoping to tag more than just `Casebook`s. It seems like I should also make `Tag` use `GenericForeignKeys`? Or perhaps a real `ForeignKey` to a `ContentNode` would be better, since I think (not quite sure) we could infer that a top-level node represents a casebook.

Reply:
I'm not a big fan of `GenericForeignKey` unless you really definitely absolutely have no idea what the connecting models might be. Here's a good writeup of the problems and some alternative architectures: https://lukeplant.me.uk/blog/posts/avoid-django-genericforeignkey/

For this case I think it'd be perfectly okay to add `ContentNode` later as a second kind of relation, sort of like [Alternative 5](https://lukeplant.me.uk/blog/posts/avoid-django-genericforeignkey/#:~:text=Alternative%205%20%2D%20multiple%20linked%20models) here. For your purposes, I recommend just pretending like we're only worrying about Casebooks for now—I think that'll help our decision-making on a lot of fronts to start simple.